<!DOCTYPE html>
<html>

  <head>
    <meta charset='utf-8' />
    <meta http-equiv="X-UA-Compatible" content="chrome=1" />
    <meta name="description" content="PALEOMIX : " />

    <link rel="stylesheet" type="text/css" media="screen" href="stylesheets/stylesheet.css">

    <title>PALEOMIX</title>
  </head>

  <body>

    <!-- HEADER -->
    <div id="header_wrap" class="outer">
        <header class="inner">
          <a id="forkme_banner" href="https://github.com/MikkelSchubert/paleomix">View on GitHub</a>

          <h1 id="project_title">PALEOMIX</h1>
          <h2 id="project_tagline"></h2>

            <section id="downloads">
              <a class="zip_download_link" href="https://github.com/MikkelSchubert/paleomix/zipball/master">Download this project as a .zip file</a>
              <a class="tar_download_link" href="https://github.com/MikkelSchubert/paleomix/tarball/master">Download this project as a tar.gz file</a>
            </section>
        </header>
    </div>

    <!-- MAIN CONTENT -->
    <div id="main_content_wrap" class="outer">
      <section id="main_content" class="inner">
        <p>Sequencing complete genomes still represented a few years ago a massive endeavor involving hundreds of researchers and thousands of sequencing machines grouped within big sequencing centres. The advent of next generation sequencing machines has completely changed this situation, by delivering up to billions of sequences within a couple of weeks. As a result, we now know the genome sequence of a variety of taxa, including non-model organisms, and even of extinct species and ancient individuals. For some groups such as humans, we even know the sequence variation segregating at the genome scale in a large number of populations around the world. The fantastic throughput of next-generation sequencing platforms has reduced sequencing costs by orders of magnitude but has at the same time increased the computational work load associated with the characterization of full genome sequences.</p>

<p>The PALEOMIX pipeline is a user-friendly package designed for UNIX and MAC systems and largely automates most of analyses related to whole genome re-sequencing. It is compatible with a full range of sequence data and performs a series of user-defined analyses, including read trimming, read collapsing, read mapping, PCR duplicate removal, SNP calling, and metagenomic profiling. For ancient DNA sequence data, PALEOMIX also supports <em>post-mortem</em> DNA damage quantification and standard mis-incorporation and fragmentation patterns. Finally, in situations where several genomes are available, PALEOMIX can reconstruct Maximum Likelihood phylogenomic trees and reveal the evolutionary phylogenetic relationships among taxa.</p>

<p>PALEOMIX has been developed by researchers from Ludovic Orlandoâ€™s group at the Centre for GeoGenetics, University of Copenhagen, Denmark. The software and related documentation is available at <a href="https://github.com/MikkelSchubert/paleomix">https://github.com/MikkelSchubert/paleomix</a></p>

<p>This work was supported by the Danish Council for Independent Research, Natural Sciences (FNU); the Danish National Research Foundation (DNRF94); a Marie-Curie Career Integration grant (FP7 CIG-293845); the Lundbeck foundation (R52-A5062). </p>

  </body>
</html>
